File: src/services/tryon.service.ts
import { prisma } from '@/lib/prisma';
import { buildUserPrompt } from './prompt-builder';
import { qualityCheck, refinePromptWithIssues } from './quality-check';

interface TryonRequest { userId: string; bodyProfileId: string; garmentIds: string[]; scene: string; pose: string; quality?: 'standard' | 'high'; sizeOverrides?: Record<string, string>; }
interface TryonResult { taskId: string; status: string; resultUrls: string[]; qualityScore?: number; }

const MAX_RETRIES = 2;

async function callImageApi(prompt: string, referenceImages: string[], quality: string): Promise<string> {
  const model = process.env.OPENAI_IMAGE_MODEL || 'gpt-image-2pro';
  const size = quality === 'high' ? '2048x2048' : '1024x1024';
  const res = await fetch(process.env.OPENAI_IMAGE_BASE_URL + '/images/generations', {
    method: 'POST',
    headers: { 'Authorization': 'Bearer ' + process.env.OPENAI_IMAGE_API_KEY, 'Content-Type': 'application/json' },
    body: JSON.stringify({ model, n: 1, size, prompt, reference_images: referenceImages.length > 0 ? referenceImages : undefined, response_format: 'b64_json' }),
  });
  if (!res.ok) throw new Error('生图 API 错误 (' + res.status + '): ' + await res.text());
  const data = await res.json();
  return data.data?.[0]?.url || '';
}

export async function executeTryon(request: TryonRequest): Promise<TryonResult> {
  const startTime = Date.now();
  const bodyProfile = await prisma.bodyProfile.findUnique({ where: { id: request.bodyProfileId } });
  if (!bodyProfile) throw new Error('体型档案不存在');
  const garments = await prisma.garment.findMany({ where: { id: { in: request.garmentIds } }, include: { sizeCharts: { orderBy: { sortOrder: 'asc' } } } });
  if (garments.length === 0) throw new Error('未找到指定服装');
  const task = await prisma.tryonTask.create({ data: { userId: request.userId, bodyProfileId: request.bodyProfileId, garmentIds: request.garmentIds, scene: request.scene, poseType: request.pose, status: 'processing' } });
  try {
    const refImages: string[] = [];
    if (bodyProfile.frontPhotoUrl) refImages.push(bodyProfile.frontPhotoUrl);
    for (const g of garments) { if (refImages.length >= 4) break; const img = g.processedImage || g.originalImage; if (img) refImages.push(img); }

    let currentPrompt = buildUserPrompt({
      bodyProfile: { heightCm: bodyProfile.heightCm ? Number(bodyProfile.heightCm) : null, weightKg: bodyProfile.weightKg ? Number(bodyProfile.weightKg) : null, shoulderWidth: bodyProfile.shoulderWidth ? Number(bodyProfile.shoulderWidth) : null, chestCirc: bodyProfile.chestCirc ? Number(bodyProfile.chestCirc) : null, waistCirc: bodyProfile.waistCirc ? Number(bodyProfile.waistCirc) : null, hipCirc: bodyProfile.hipCirc ? Number(bodyProfile.hipCirc) : null, bodyType: bodyProfile.bodyType, bodyDescription: bodyProfile.bodyDescription },
      garments: garments.map(g => {
        const ov = request.sizeOverrides?.[g.id];
        let sl = g.sizeCharts[0]?.sizeLabel;
        let sm: Record<string, number> | undefined;
        if (ov && g.sizeCharts.length > 0) {
          const t = g.sizeCharts.find(s => s.sizeLabel === ov);
          if (t) { sl = t.sizeLabel; sm = {}; if (t.chest) sm.chest = Number(t.chest); if (t.shoulder) sm.shoulder = Number(t.shoulder); if (t.waistCirc) sm.waistCirc = Number(t.waistCirc); if (t.hipCirc) sm.hipCirc = Number(t.hipCirc); if (t.sleeveLength) sm.sleeveLength = Number(t.sleeveLength); if (t.totalLength) sm.totalLength = Number(t.totalLength); if (t.inseam) sm.inseam = Number(t.inseam); }
        }
        return { category: g.category, name: g.name, colorPrimary: g.colorPrimary, material: g.material, fitType: g.fitType, pattern: g.pattern, aiDescription: g.aiDescription, sizeLabel: sl, sizeMeasurements: sm };
      }),
      scene: request.scene, pose: request.pose, quality: request.quality,
    });

    let resultUrl = ''; let qualityScore = 7; let retryCount = 0;
    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
      resultUrl = await callImageApi(currentPrompt, refImages, request.quality || 'standard');
      if (!resultUrl) throw new Error('生图 API 未返回结果');
      if (bodyProfile.frontPhotoUrl) {
        const check = await qualityCheck(bodyProfile.frontPhotoUrl, resultUrl);
        qualityScore = check.score;
        if (check.pass || attempt === MAX_RETRIES) { retryCount = attempt; break; }
        currentPrompt = refinePromptWithIssues(currentPrompt, check.issues);
        retryCount = attempt + 1;
      } else { break; }
    }

    const pt = Date.now() - startTime;
    await prisma.tryonTask.update({ where: { id: task.id }, data: { status: 'completed', resultUrls: resultUrl ? [resultUrl] : [], promptUsed: currentPrompt, apiModel: process.env.OPENAI_IMAGE_MODEL || 'gpt-image-2pro', apiCallsCount: retryCount + 1, processingTimeMs: pt, qualityScore, retryCount, completedAt: new Date() } });
    return { taskId: task.id, status: 'completed', resultUrls: resultUrl ? [resultUrl] : [], qualityScore };
  } catch (error) {
    await prisma.tryonTask.update({ where: { id: task.id }, data: { status: 'failed', errorMessage: error instanceof Error ? error.message : '未知错误', processingTimeMs: Date.now() - startTime } });
    throw error;
  }
}

export async function getTaskStatus(taskId: string, userId: string) {
  const task = await prisma.tryonTask.findFirst({ where: { id: taskId, userId } });
  if (!task) return null;
  return { taskId: task.id, status: task.status, resultUrls: task.resultUrls, qualityScore: task.qualityScore ? Number(task.qualityScore) : null, retryCount: task.retryCount, processingTimeMs: task.processingTimeMs, errorMessage: task.errorMessage, createdAt: task.createdAt, completedAt: task.completedAt };
}

File: backend/app/services/tryon_service.py
import time
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from app.models.body_profile import BodyProfile
from app.models.garment import Garment, SizeChart
from app.models.tryon_task import TryonTask
from app.services.ai_service import call_image_api, quality_check

MAX_RETRIES = 2

def build_user_prompt(bp, garments, scene, pose):
    lines = ["## 用户信息"]
    if bp.height_cm: lines.append("- 身高：" + str(float(bp.height_cm)) + "cm")
    if bp.weight_kg: lines.append("- 体重：" + str(float(bp.weight_kg)) + "kg")
    m = []
    if bp.shoulder_width: m.append("肩宽" + str(float(bp.shoulder_width)) + "cm")
    if bp.chest_circ: m.append("胸围" + str(float(bp.chest_circ)) + "cm")
    if bp.waist_circ: m.append("腰围" + str(float(bp.waist_circ)) + "cm")
    if bp.hip_circ: m.append("臀围" + str(float(bp.hip_circ)) + "cm")
    if m: lines.append("- 体型数据：" + "，".join(m))
    if bp.body_type: lines.append("- 体型分类：" + bp.body_type)
    if bp.body_description: lines.append("- 体型特征：" + bp.body_description)

    lines.append("\n## 穿搭要求\n请让图中人物穿上以下服装：")
    cat_names = {"top": "上衣", "bottom": "裤装", "dress": "连衣裙", "shoes": "鞋履", "accessory": "配饰", "bag": "包袋"}
    for i, g in enumerate(garments):
        parts = ["- [附件" + str(i+1) + "] " + cat_names.get(g.category, g.category) + "：" + (g.name or "未命名")]
        d = []
        if g.color_primary: d.append("颜色：" + g.color_primary)
        if g.material: d.append("材质：" + g.material)
        if g.fit_type: d.append("版型：" + g.fit_type)
        if g.pattern: d.append("图案：" + g.pattern)
        if d: parts.append("  · " + "，".join(d))
        if g.ai_description: parts.append("  · 描述：" + g.ai_description)
        lines.append("\n".join(parts))

    scene_descs = {"studio": "专业摄影棚，白色无缝背景，均匀柔光", "street": "繁华都市街头，自然日光", "indoor": "简约现代风格客厅，柔和灯光", "outdoor": "阳光明媚的户外花园，绿色植物背景"}
    pose_descs = {"front_standing": "正面自然站立，双手自然下垂，目视镜头", "side_45": "身体微侧45度，展示服装侧面轮廓", "walking": "走路中的自然姿态，步伐轻盈", "sitting": "优雅地坐在高脚椅上，双腿交叉"}
    lines.append("\n## 场景与姿势")
    lines.append("- 背景场景：" + scene_descs.get(scene, scene_descs["studio"]))
    lines.append("- 模特姿势：" + pose_descs.get(pose, pose_descs["front_standing"]))
    lines.append("\n## 质量要求\n- 风格：专业时尚摄影风格\n- 特别注意：保持面部特征完全一致，不要改变发型，服装自然穿着")
    return "\n".join(lines)

async def execute_tryon(db: AsyncSession, user_id: str, bp, garment_ids: list, scene: str, pose: str, quality: str, size_overrides: dict = None):
    result = await db.execute(select(Garment).where(Garment.id.in_(garment_ids)).options(selectinload(Garment.size_charts)))
    garments = list(result.scalars().all())
    if not garments:
        raise Exception("未找到指定服装")

    task = TryonTask(user_id=user_id, body_profile_id=bp.id, garment_ids=garment_ids, scene=scene, pose_type=pose, status="processing")
    db.add(task)
    await db.commit()
    await db.refresh(task)

    start = time.time()
    try:
        ref_images = []
        if bp.front_photo_url:
            ref_images.append(bp.front_photo_url)
        for g in garments:
            if len(ref_images) >= 4:
                break
            img = g.processed_image or g.original_image
            if img:
                ref_images.append(img)

        prompt = build_user_prompt(bp, garments, scene, pose)
        result_url = await call_image_api(prompt, ref_images, quality)
        if not result_url:
            raise Exception("生图 API 未返回结果")

        quality_score = 7
        if bp.front_photo_url:
            check = await quality_check(bp.front_photo_url, result_url)
            quality_score = check["score"]
            if not check["pass"]:
                for attempt in range(MAX_RETRIES):
                    prompt = prompt + "\n\n## 优化要求\n" + "；".join(check.get("issues", []))
                    result_url = await call_image_api(prompt, ref_images, quality)
                    check = await quality_check(bp.front_photo_url, result_url)
                    quality_score = check["score"]
                    if check["pass"]:
                        break

        pt = int((time.time() - start) * 1000)
        task.status = "completed"
        task.result_urls = [result_url] if result_url else []
        task.prompt_used = prompt
        task.api_model = "gpt-image-2pro"
        task.api_calls_count = 1 + MAX_RETRIES
        task.processing_time_ms = pt
        task.quality_score = quality_score
        task.completed_at = time.time()
        await db.commit()
        return {"task_id": task.id, "status": "completed", "result_urls": task.result_urls, "quality_score": quality_score}
    except Exception as e:
        task.status = "failed"
        task.error_message = str(e)
        task.processing_time_ms = int((time.time() - start) * 1000)
        await db.commit()
        raise
